upcloud/service/retry.go

```diff
@@ -23,22 +28,46 @@ func fillDefaults(c *retryConfig) *retryConfig {
 	return c
 }
 
-func retry[T any](ctx context.Context, operation func(int, context.Context) (*T, error), config *retryConfig) (*T, error) {
+func shouldRetryOnError(err error, retryOnErrorCount int) bool {
+	if retryOnErrorCount >= maxRetriesOnError {
+		return false
+	}
+
+	var ucErr *upcloud.Problem
+	if errors.As(err, &ucErr) && ucErr.Status >= 500 {
+		return true
+	}
+
+	return false
+}
+
+func Retry[T any](ctx context.Context, operation func(int, context.Context) (*T, error), config *RetryConfig) (*T, error) {
```

Comment:
Or should we keep this as private function and move all wait-until loops from TF provider and upctl to Go SDK? 🤔